### PR TITLE
Remove symlink when removing vmgroup without "--remove-volume" option.

### LINKED ISSUE #1452: Recreate vmgroup with the same name after "vmgroup rm " cannot create vmgroup correctly
1. create a vmgroup "test-tenant1" and then create a volume "tenant1-vol1"
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=test-tenant1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0 --default-datastore=sharedVmfs-0
vmgroup 'test-tenant1' is created. Do not forget to run 'vmgroup vm add' to add vm to vmgroup.

root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=tenant1-vol1
tenant1-vol1
root@sc-rdops-vm02-dhcp-52-237:~# 
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      tenant1-vol1@sharedVmfs-0

```

2. remove the vmgroup "test-tenant1".
```
root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup vm rm --name=test-tenant1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0
vmgroup vm rm succeeded
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup rm --name=test-tenant1
vmgroup rm succeeded

[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup ls
Uuid                                  Name      Description                Default_datastore  VM_list  
------------------------------------  --------  -------------------------  -----------------  -------  
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default vmgroup  _VM_DS           
```

3. check the dockvol path of "test_tenant1" 

```
[root@sc-rdops-vm17-dhcp-3-236:~] ls -al /vmfs/volumes/sharedVmfs-0/dockvols/
total 1064
drwxr-xr-x    1 root     root           840 Jun 21 15:58 .
drwxr-xr-t    1 root     root          3080 Jun 20 23:22 ..
drwxr-xr-x    1 root     root           700 Jun 21 15:58 4db87e90-cfc3-4291-ac99-310296cf5c71
drwxr-xr-x    1 root     root           280 Jun 21 06:18 631407d5-6529-4a52-8624-a99787c98f2a
drwxr-xr-x    1 root     root          1540 Jun 21 06:18 7a1944c7-7632-48bc-a3f7-b54a1c240392
lrwxrwxrwx    1 root     root            72 Jun 21 15:58 test-tenant1 -> /vmfs/volumes/sharedVmfs-0/dockvols/4db87e90-cfc3-4291-ac99-310296cf5c71      ---> still point to the dockvol path of "test-tenant1"
```

4. recreate "test_tenant1"
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=test-tenant1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0 --default-datastore=sharedVmfs-0
vmgroup 'test-tenant1' is created. Do not forget to run 'vmgroup vm add' to add vm to vmgroup.
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup ls
Uuid                                  Name          Description                Default_datastore  VM_list                    
------------------------------------  ------------  -------------------------  -----------------  -------------------------  
11111111-1111-1111-1111-111111111111  _DEFAULT      This is a default vmgroup  _VM_DS                                        
6ed84780-3597-45bb-9d63-164980a945a6  test-tenant1                             sharedVmfs-0       ubuntu-VM0.0,ubuntu-VM1.0     --> "test-tenant1" has the new UUID after recreate

[root@sc-rdops-vm17-dhcp-3-236:~] ls -al /vmfs/volumes/sharedVmfs-0/dockvols/
total 1064
drwxr-xr-x    1 root     root           840 Jun 21 15:58 .
drwxr-xr-t    1 root     root          3080 Jun 20 23:22 ..
drwxr-xr-x    1 root     root           700 Jun 21 15:58 4db87e90-cfc3-4291-ac99-310296cf5c71
drwxr-xr-x    1 root     root           280 Jun 21 06:18 631407d5-6529-4a52-8624-a99787c98f2a
drwxr-xr-x    1 root     root          1540 Jun 21 06:18 7a1944c7-7632-48bc-a3f7-b54a1c240392
lrwxrwxrwx    1 root     root            72 Jun 21 15:58 test-tenant1 -> /vmfs/volumes/sharedVmfs-0/dockvols/4db87e90-cfc3-4291-ac99-310296cf5c71   ---> symlink still point to the old UUID, which is wrong


```

5. Try to create the volume, and volume does not show up  when run ```docker volume ls```
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=tenant1-vol2
tenant1-vol2
root@sc-rdops-vm02-dhcp-52-237:~# 
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME

```

### PR DESCRIPTION
**Pull request by lipingxue**
Fixed #1452
Found this issue when I work on my automation test. 
Right now, the following tests are done manually. Later another PR for my automation test will cover this test too.

1. create a vmgroup "test_vmgroup_1" and add two VMs to it
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=test_vmgroup_1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0 --default-datastore=sharedVmfs-0
vmgroup 'test_vmgroup_1' is created. Do not forget to run 'vmgroup vm add' to add vm to vmgroup.
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup ls
Uuid                                  Name            Description                Default_datastore  VM_list                   
------------------------------------  --------------  -------------------------  -----------------  -------------------------  
11111111-1111-1111-1111-111111111111  _DEFAULT        This is a default vmgroup  _VM_DS                                       
46d0c05d-f164-484c-8976-7420fad35e62  test_vmgroup_1                             sharedVmfs-0       ubuntu-VM0.0,ubuntu-VM1.0  
```

2. create a volume "vmgroup1-vol1" for vmgroup "test_vmgroup1"
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=vmgroup1-vol1
vmgroup1-vol1
root@sc-rdops-vm02-dhcp-52-237:~# 
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      vmgroup1-vol1@sharedVmfs-0
```

3.  check the symlink for vmgroup "test_vmgroup_1" which points to the right path /vmfs/volume/datastore_name/tenant_uuid
```
[root@sc-rdops-vm17-dhcp-3-236:~] ls -al /vmfs/volumes/sharedVmfs-0/dockvols/
total 1048
drwxr-xr-x    1 root     root           560 Jun 21 22:20 .
drwxr-xr-t    1 root     root          3080 Jun 20 23:22 ..
drwxr-xr-x    1 root     root           700 Jun 21 22:20 46d0c05d-f164-484c-8976-7420fad35e62
lrwxrwxrwx    1 root     root            72 Jun 21 22:20 test_vmgroup_1 -> /vmfs/volumes/sharedVmfs-0/dockvols/46d0c05d-f164-484c-8976-7420fad35e62
```

4. remove VMs from vmgroup "test_vmgroup_1" and then remove the vmgroup without "--remove-volume" option
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup vm rm --name=test_vmgroup_1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0 
vmgroup vm rm succeeded
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup rm --name=test_vmgroup_1
vmgroup rm succeeded
```

5. run ```vmdkops_admin volume ls```, can see this volume, but does not belong to any vmgroup
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume         Datastore     VMGroup  Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By    Created Date              
-------------  ------------  -------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  ------------  ------------------------  
vmgroup1-vol1  sharedVmfs-0  N/A      100MB     13MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM0.0  Wed Jun 21 22:20:05 2017  
```

6. check the symlink, symlink for this vmgroup has been removed
```
[root@sc-rdops-vm17-dhcp-3-236:~] ls -al /vmfs/volumes/sharedVmfs-0/dockvols/
total 1040
drwxr-xr-x    1 root     root           420 Jun 21 22:21 .
drwxr-xr-t    1 root     root          3080 Jun 20 23:22 ..
drwxr-xr-x    1 root     root           700 Jun 21 22:20 46d0c05d-f164-484c-8976-7420fad35e62
```

7. recreate the vmgroup with the same name "test_vmgroup_1"
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=test_vmgroup_1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0 --default-datastore=sharedVmfs-0
vmgroup 'test_vmgroup_1' is created. Do not forget to run 'vmgroup vm add' to add vm to vmgroup.
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup ls
Uuid                                  Name            Description                Default_datastore  VM_list                   
------------------------------------  --------------  -------------------------  -----------------  -------------------------  
11111111-1111-1111-1111-111111111111  _DEFAULT        This is a default vmgroup  _VM_DS                                       
80bcaf71-6b96-4968-b03c-0b134475a687  test_vmgroup_1                             sharedVmfs-0       ubuntu-VM0.0,ubuntu-VM1.0  
```

8. create a new volume, ```docker volume ls``` can list the right volume
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=new-vmgroup1-vol1
new-vmgroup1-vol1
root@sc-rdops-vm02-dhcp-52-237:~# 
root@sc-rdops-vm02-dhcp-52-237:~# 
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      new-vmgroup1-vol1@sharedVmfs-0
``` 

9. run ```vmdkops_admin volume ls```, and can see both volumes
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume             Datastore     VMGroup         Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By    Created Date              
-----------------  ------------  --------------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  ------------  ------------------------  
vmgroup1-vol1      sharedVmfs-0  N/A             100MB     13MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM0.0  Wed Jun 21 22:20:05 2017  
new-vmgroup1-vol1  sharedVmfs-0  test_vmgroup_1  100MB     13MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM0.0  Wed Jun 21 22:22:50 2017  
```

10. check the symlink, symlink for this vmgroup link to the path /vmfs/volumes/datastore_name/new_tenant_uuid

```
[root@sc-rdops-vm17-dhcp-3-236:~] ls -al /vmfs/volumes/sharedVmfs-0/dockvols/
total 1056
drwxr-xr-x    1 root     root           700 Jun 21 22:22 .
drwxr-xr-t    1 root     root          3080 Jun 20 23:22 ..
drwxr-xr-x    1 root     root           700 Jun 21 22:20 46d0c05d-f164-484c-8976-7420fad35e62
drwxr-xr-x    1 root     root           700 Jun 21 22:22 80bcaf71-6b96-4968-b03c-0b134475a687
lrwxrwxrwx    1 root     root            72 Jun 21 22:22 test_vmgroup_1 -> /vmfs/volumes/sharedVmfs-0/dockvols/80bcaf71-6b96-4968-b03c-0b134475a687
```

### CODE REVIEW

File: esx_service/utils/auth_data.py [hunk 1]
```diff
@@ -1246,6 +1259,10 @@ def remove_tenant(self, tenant_id, remove_volumes):
             error_msg = self.__remove_volumes_for_tenant(tenant_id)
             if error_msg:
                 return error_msg
+        else:
+            error_msg = self.remove_symlink_for_tenant(tenant_id)
```
Comment (govint) [inline esx_service/utils/auth_data.py]: Can this new function be used in __remove_volumes_for_tenant() vs. having the symlink removal code in two places once in this function and in __remove_volumes_for_tenant() as well.

File: esx_service/utils/auth_data.py [hunk 2]
```diff
@@ -1157,6 +1157,27 @@ def remove_volumes_from_volumes_table(self, tenant_id):
...
+    def remove_symlink_for_tenant(self, tenant_id):
+        """
+            Delete the symlink /vmfs/volume/datastore_name/tenant_name
+            which point to /vmfs/volumes/datastore_name/tenant_uuid
+        """
+        err, tenant_name = self.get_tenant_name(tenant_id)
+        if err:
+            return err
+        VOL_RM_LOG_PREFIX = "Tenant <name> %s removal: "
+        for (datastore, _, path) in vmdk_utils.get_datastores():
```
Comment (govint) [inline esx_service/utils/auth_data.py]: Does this have to be a loop over all datastores while wanting to remove only the symlink for a specific datastore? Can this loop over instead on just the datastores assigned to the vmgroup.

What happens if there are multiple datastores assigned to a single tenant, that is also a test case and should also be verified.

File: esx_service/utils/auth_data.py (hunk 1 above)
Comment (lipingxue) [inline esx_service/utils/auth_data.py]: Function ```remove_symlink_for_tenant``` is reused. So what is your question?

File: esx_service/utils/auth_data.py (hunk 2 above)
Comment (lipingxue) [inline esx_service/utils/auth_data.py]: I don't understand the first part. What is the problem to loop over all datastores? I don't expect to have many datastores. How can we figure out the datastores assigned to the vmgroup without loop through all datastores? Could you explain more?

For the second part, yes,  I can add a test case for that.

File: esx_service/utils/auth_data.py (hunk 1 above)
Comment (govint) [inline esx_service/utils/auth_data.py]: Correct, sorry i did miss the change in remove_volumes_for_tenant(). Ok.

**Comment by lipingxue** [comment]
Add test when vmgroup has volumes in more than one datastore 
1. create vmgroup and add VMs
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=test_vmgroup_1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0 --default-datastore=sharedVmfs-0
vmgroup 'test_vmgroup_1' is created. Do not forget to run 'vmgroup vm add' to add vm to vmgroup.

```
2. create two volumes, one is on the default_datastore, and the other is on vsanDatastore
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=vmgroup-vol1
vmgroup-vol1

root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=vmgroup-vol2@vsanDatastore
vmgroup-vol2@vsanDatastore
root@sc-rdops-vm02-dhcp-52-237:~# 
root@sc-rdops-vm02-dhcp-52-237:~# 
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      vmgroup-vol1@sharedVmfs-0
vsphere:latest      vmgroup-vol2@vsanDatastore

```
3. check the symlink
```
[root@sc-rdops-vm17-dhcp-3-236:~] ls /vmfs/volumes/sharedVmfs-0/dockvols/ -al
total 1048
drwxr-xr-x    1 root     root           560 Jun 22 16:54 .
drwxr-xr-t    1 root     root          3080 Jun 22 16:52 ..
drwxr-xr-x    1 root     root           700 Jun 22 16:54 af6c491c-eba4-41c3-bc5e-4dfc5223acf2
lrwxrwxrwx    1 root     root            72 Jun 22 16:54 test_vmgroup_1 -> /vmfs/volumes/sharedVmfs-0/dockvols/af6c491c-eba4-41c3-bc5e-4dfc5223acf2
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] ls /vmfs/volumes/vsanDatastore/dockvols/ -al
total 790552
drwxr-xr-t    1 root     root          1680 Jun 22 16:55 .
drwxr-xr-x    1 root     root           512 Jun 22 16:56 ..
-r--------    1 root     root       1441792 Jun 20 23:22 .fbb.sf
-r--------    1 root     root     267026432 Jun 20 23:22 .fdc.sf
-r--------    1 root     root       1179648 Jun 20 23:22 .pb2.sf
-r--------    1 root     root     268435456 Jun 20 23:22 .pbc.sf
-r--------    1 root     root     262733824 Jun 20 23:22 .sbc.sf
drwx------    1 root     root           280 Jun 20 23:22 .sdd.sf
-r--------    1 root     root       4194304 Jun 20 23:22 .vh.sf
drwxr-xr-x    1 root     root           700 Jun 22 16:55 af6c491c-eba4-41c3-bc5e-4dfc5223acf2
drwxr-xr-x    1 root     root           280 Jun 21 22:28 policies
lrwxrwxrwx    1 root     root            73 Jun 22 16:55 test_vmgroup_1 -> /vmfs/volumes/vsanDatastore/dockvols/af6c491c-eba4-41c3-bc5e-4dfc5223acf2

```
4.  remove VMs from the vmgroup and then remove the vmgroup
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup vm rm --name=test_vmgroup_1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0
vmgroup vm rm succeeded
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup rm --name=test_vmgroup_1
vmgroup rm succeeded
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume        Datastore      VMGroup  Capacity  Used  Filesystem  Policy          Disk Format  Attached-to  Access      Attach-as               Created By    Created Date              
------------  -------------  -------  --------  ----  ----------  --------------  -----------  -----------  ----------  ----------------------  ------------  ------------------------  
vmgroup-vol1  sharedVmfs-0   N/A      100MB     13MB  ext4        N/A             thin         detached     read-write  independent_persistent  ubuntu-VM0.0  Thu Jun 22 16:54:09 2017  
vmgroup-vol2  vsanDatastore  N/A      100MB     80MB  ext4        [VSAN default]  thin         detached     read-write  independent_persistent  ubuntu-VM0.0  Thu Jun 22 16:55:47 2017  

```
5. check the symlink, symlink are removed on both datastores
```
[root@sc-rdops-vm17-dhcp-3-236:~] ls /vmfs/volumes/sharedVmfs-0/dockvols/ -al
total 1040
drwxr-xr-x    1 root     root           420 Jun 22 16:58 .
drwxr-xr-t    1 root     root          3080 Jun 22 16:52 ..
drwxr-xr-x    1 root     root           700 Jun 22 16:54 af6c491c-eba4-41c3-bc5e-4dfc5223acf2
[root@sc-rdops-vm17-dhcp-3-236:~] ls /vmfs/volumes/vsanDatastore/dockvols/ -al
total 790544
drwxr-xr-t    1 root     root          1540 Jun 22 16:58 .
drwxr-xr-x    1 root     root           512 Jun 22 16:58 ..
-r--------    1 root     root       1441792 Jun 20 23:22 .fbb.sf
-r--------    1 root     root     267026432 Jun 20 23:22 .fdc.sf
-r--------    1 root     root       1179648 Jun 20 23:22 .pb2.sf
-r--------    1 root     root     268435456 Jun 20 23:22 .pbc.sf
-r--------    1 root     root     262733824 Jun 20 23:22 .sbc.sf
drwx------    1 root     root           280 Jun 20 23:22 .sdd.sf
-r--------    1 root     root       4194304 Jun 20 23:22 .vh.sf
drwxr-xr-x    1 root     root           700 Jun 22 16:55 af6c491c-eba4-41c3-bc5e-4dfc5223acf2
drwxr-xr-x    1 root     root           280 Jun 21 22:28 policies

```

**Comment by shuklanirdesh82** [comment]
@lipingxue Do we have an automation test for this one?

**Pull request merged**